src/Sema.zig

```diff
@@ -12637,6 +12670,180 @@ fn zirReify(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air.I
+                    .{ field_name, err },
+                );
+            };
+
+            if (field_index >= fields_len) {
+                return sema.fail(
+                    block,
+                    src,
+                    "tuple field {} exceeds tuple field count",
+                    .{field_index},
+                );
+            }
+
+            const gop = used_fields.getOrPutAssumeCapacity(field_index);
+            if (gop.found_existing) {
+                // TODO: better source location
+                return sema.fail(block, src, "duplicate tuple field {}", .{field_index});
+            }
+
+            const default_val = if (default_value_val.optionalValue()) |opt_val| blk: {
+                const payload_val = if (opt_val.pointerDecl()) |opt_decl|
+                    opt_decl.val
+                else
+                    opt_val;
+                break :blk try payload_val.copy(sema.arena);
+            } else Value.initTag(.unreachable_value);
+
+            var buffer: Value.ToTypeBuffer = undefined;
+            types[field_index] = try field_type_val.toType(&buffer).copy(sema.arena);
+            values[field_index] = default_val;
+        }
+    }
+
+    const ty = try Type.Tag.tuple.create(sema.arena, .{
+        .types = types,
+        .values = values,
+    });
+    return sema.addType(ty);
+}
+
+fn reifyStruct(
+    sema: *Sema,
+    block: *Block,
+    inst: Zir.Inst.Index,
+    src: LazySrcLoc,
+    layout_val: Value,
+    fields_val: Value,
+) CompileError!Air.Inst.Ref {
+    var new_decl_arena = std.heap.ArenaAllocator.init(sema.gpa);
+    errdefer new_decl_arena.deinit();
+    const new_decl_arena_allocator = new_decl_arena.allocator();
+
+    const struct_obj = try new_decl_arena_allocator.create(Module.Struct);
+    const struct_ty = try Type.Tag.@"struct".create(new_decl_arena_allocator, struct_obj);
+    const new_struct_val = try Value.Tag.ty.create(new_decl_arena_allocator, struct_ty);
+    const type_name = try sema.createTypeName(block, .anon);
+    const new_decl = try sema.mod.createAnonymousDeclNamed(block, .{
+        .ty = Type.type,
+        .val = new_struct_val,
+    }, type_name);
+    new_decl.owns_tv = true;
+    errdefer sema.mod.abortAnonDecl(new_decl);
+    struct_obj.* = .{
+        .owner_decl = new_decl,
+        .fields = .{},
+        .node_offset = src.node_offset,
+        .zir_index = inst,
+        .layout = layout_val.toEnum(std.builtin.Type.ContainerLayout),
+        .status = .have_layout,
+        .known_non_opv = undefined,
+        .namespace = .{
+            .parent = block.namespace,
+            .ty = struct_ty,
+            .file_scope = block.getFileScope(),
+        },
+    };
+
+    // Fields
+    const slice_val = fields_val.castTag(.slice).?.data;
+    const decl = slice_val.ptr.pointerDecl().?;
+    try sema.ensureDeclAnalyzed(decl);
+    const fields_len = try sema.usizeCast(block, src, decl.ty.arrayLen());
```

Comment:
same deal here, use the slice methods of Value. Also if there end up being no fields, then I think you can just return `Type.initTag(.empty_struct)`, no Decl required. There's even an `Air.Inst.Ref` tag for it. You can straight up return that from the function.